pandas/io/parsers.py

```diff
@@ -1894,23 +1899,52 @@ def __init__(self, src, **kwds):
 
         ParserBase.__init__(self, kwds)
 
-        encoding = kwds.get("encoding")
-
-        # parsers.TextReader doesn't support compression dicts
-        if isinstance(kwds.get("compression"), dict):
-            kwds["compression"] = kwds["compression"]["method"]
-
-        if kwds.get("compression") is None and encoding:
-            if isinstance(src, str):
-                src = open(src, "rb")
-                self.handles.append(src)
-
-            # Handle the file object with universal line mode enabled.
-            # We will handle the newline character ourselves later on.
-            if hasattr(src, "read") and not hasattr(src, "encoding"):
-                src = TextIOWrapper(src, encoding=encoding, newline="")
-
-            kwds["encoding"] = "utf-8"
+        src = stringify_path(src)
+        if (
+            kwds.get("memory_map", False)
+            and isinstance(src, str)
+            and get_compression_method(kwds.get("compression", None))[0] is None
+        ):
+            # memory-mapped files are directly handled by the TextReader.
+            # compression and file-object were never supported by it.
+            self.source = src
+            self.handles = []
+        else:
+            if kwds.get("memory_map", False):
+                if not isinstance(src, str):
+                    warnings.warn(
+                        "read_csv does support only string-like objects with "
+                        + "memory_map=True and engine='c'. Falling back to "
```

Comment:
why are these warnings? we can't directly raise?

Reply:
that branch sets `memory_map` to False, so a warning is sufficient. But we could also raise an error, if we want the user to force to change their code.

Reply:
unless this breaks current tests (and so we would need some sort of deprecation), i would prefer to be explict on errors rather than fallback, esp for a case like this as memory_map requires a file.

Reply:
There are some tests but I already adjusted them a little bit (see the changes in pandas/tests/io/parser/test_textreader.py).

Some tests in pandas/tests/io/parser/test_textreader.py tried to test `memory_map=True ` with a file object. That never reached the memory_map branch in the c-engine (it only ever worked for strings). I'm not familiar enough with the c/cython code to create/manage mmap to judge whether it could also use a file handle. Outsourcing the mmap part to `get_handle` would be my preferred solution, but that lead to two errors that I could narrow done (using an already closed file handle and a segfault).

The python engine supports some type of file handles with `memory_map=True` but not all of them (`pd.read_csv("test.csv", compression="gzip", memory_map=True, engine="python")` fails for example)

Reply:
Indeed, `memory_map` is not considered a fully supported Python engine function.

I agree that explicit should be preferred, but if it's going to overly complicate this PR, we can defer to follow-up.

Reply:
changed to raise `ValueError`